geps/gep-2643/index.md

```diff
@@ -0,0 +1,514 @@
+// Core: The listener CAN be of type Passthrough
+// Extended: The listener CAN be of type Terminate 
+type TLSRouteSpec struct {
+  // Hostnames defines a set of SNI hostnames that should match against the
+  // SNI attribute of TLS ClientHello message in TLS handshake. This matches
+  // the RFC 1123 definition of a hostname with 2 notable exceptions:
+  //
+  // 1. IPs are not allowed in SNI hostnames per RFC 6066.
+  // 2. A hostname may be prefixed with a wildcard label (`*.`). The wildcard
+  //    label must appear by itself as the first label.
+  //
+  // <gateway:util:excludeFromCRD>
+  // If a hostname is specified by both the Listener and TLSRoute, there
+  // must be at least one intersecting hostname for the TLSRoute to be
+  // attached to the Listener. For example:
+  //
+  // * A Listener with `test.example.com` as the hostname matches TLSRoutes
+  //   that have specified at least one of `test.example.com` or
+  //   `*.example.com`.
+  // * A Listener with `*.example.com` as the hostname matches TLSRoutes
+  //   that have specified at least one hostname that matches the Listener
+  //   hostname. For example, `test.example.com` and `*.example.com` would both
+  //   match. On the other hand, `example.com` and `test.example.net` would not
+  //   match.
+  // * A listener with `something.example.com` as the hostname matches a 
+  //   TLSRoute with hostname `*.example.com`.
+  //
+  // If both the Listener and TLSRoute have specified hostnames, any
+  // TLSRoute hostnames that do not match any Listener hostname MUST be
+  // ignored. For example, if a Listener specified `*.example.com`, and the
+  // TLSRoute specified `test.example.com` and `test.example.net`,
+  // `test.example.net` must not be considered for a match.
+  //
+  // If both the Listener and TLSRoute have specified hostnames, and none
+  // match with the criteria above, then the TLSRoute is not accepted for that 
+  // Listener. If the TLSRoute does not match any Listener on the parent, the 
+  // implementation must raise an 'Accepted' Condition with a status of
+  // `False` in the corresponding RouteParentStatus.
+  // 
+  // A Listener MUST be of type TLS when a TLSRoute attaches to it. The 
+  // implementation MUST raise an 'Accepted' Condition with a status of
+  // `False` in the corresponding RouteParentStatus with the reason 
+  // of "UnsupportedValue" in case a Listener of the wrong type is used.
+  // Core: A TLS listener CAN have mode Passthrough
+  // Extended: A TLS listener CAN have mode Terminate 
+  // </gateway:util:excludeFromCRD>
+  // +required
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=16
+  Hostnames []Hostname `json:"hostnames,omitempty"`
+  // Rules is a list of TLS matchers and actions.
+  //
+  // +required
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=1
+  Rules []TLSRouteRule `json:"rules,omitempty"`
```

Comment:
No one has mentioned timeouts for this route type?  Why not include them? Won't the client want to know if the connection timed out?  

Also, if frontend tls termination is to be allowed, what about the backend being able to perform header or path matching?  We'd need more rules to express that. (But I do prefer that we don't allow frontend tls termination on TLSRoute for this reason and for keeping TLSRoute single-purpose.)